File: 0134-gas-station/0134-gas-station.py
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        if sum(cost) > sum(gas):
            return -1
        # else: a solution exists
        n = len(gas)
        gas_tank = 0
        res = 0
        for i in range(n):
            gas_tank += (-cost[i] + gas[i])
            if gas_tank < 0:  # res can't be the initial start
                gas_tank = 0  # empty gas tank
                res = i + 1  # test if an issue occur starting from i + 1
        return res

    # TLE O(n^2) ...
    """
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        n = len(gas)
        L = [-cost[i]+gas[i+1] for i in range(n-1)] + [-cost[n-1]+gas[0]]
        if sum(L) < 0:
            return -1
        for i in range(n):
            if gas[i] >= cost[i]:  # ok candidate
                res = gas[i] + L[i]
                j = (i + 1) % n
                while (res >= 0) and (j != i) and (res >= cost[j]):
                    res += L[j]
                    j += 1
                    j %= n
                if j == i:
                    return i
        return -1
    """

File: 0134-gas-station/0134-gas-station.rs
impl Solution {
    pub fn can_complete_circuit(gas: Vec<i32>, cost: Vec<i32>) -> i32 {
        let cost_sum: i32 = cost.iter().sum();
        let gas_sum: i32 = gas.iter().sum();
        if (cost_sum > gas_sum) {
            return -1;
        }
        // else: a solution exists
        let n: i32 = gas.len() as i32;
        let mut gas_tank: i32 = 0;
        let mut res: i32 = 0;
        for i in 0..n as usize {
            gas_tank += (-cost[i] + gas[i]);
            if (gas_tank < 0) {  // res can't be the initial start
                gas_tank = 0;  // empty gas tank
                res = (i as i32) + 1;  // test if an issue occur starting from i + 1
            }
        }
        return res;
    }
}
